core/src/packets/icmp/v4/redirect.rs

```diff
@@ -0,0 +1,267 @@
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+* SPDX-License-Identifier: Apache-2.0
+*/
+
+use crate::packets::icmp::v4::{Icmpv4, Icmpv4Message, Icmpv4Packet, Icmpv4Type, Icmpv4Types};
+use crate::packets::ip::v4::IPV4_MIN_MTU;
+use crate::packets::{Internal, Packet};
+use crate::SizeOf;
+use failure::Fallible;
+use std::fmt;
+use std::net::Ipv4Addr;
+use std::ptr::NonNull;
+
+/// Redirect Message defined in [IETF RFC 792].
+///
+/// ```
+///  0                   1                   2                   3
+///  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |     Type      |     Code      |          Checksum             |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |                 Gateway Internet Address                      |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |      Internet Header + 64 bits of Original Data Datagram      |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// ```
+/// *Type*: Identifier for redirect message = 5
+/// *Gateway Internet Address*: The ip address of the redirected gateway
+///
+/// [IETF RFC 792]: https://tools.ietf.org/html/rfc792
+#[derive(Icmpv4Packet)]
+pub struct Redirect {
+    icmp: Icmpv4,
+    body: NonNull<RedirectBody>,
+}
+
+impl Redirect {
+    #[inline]
+    fn body(&self) -> &RedirectBody {
+        unsafe { self.body.as_ref() }
+    }
+
+    #[inline]
+    fn body_mut(&mut self) -> &mut RedirectBody {
+        unsafe { self.body.as_mut() }
+    }
+
+    /// Returns the gateway ip address
+    #[inline]
+    pub fn gateway(&self) -> Ipv4Addr {
+        self.body().gateway
+    }
+
+    /// Sets the gateway ip address
+    #[inline]
+    pub fn set_gateway(&mut self, gateway: Ipv4Addr) {
+        self.body_mut().gateway = gateway
+    }
+
+    #[inline]
+    fn data_offset(&self) -> usize {
+        self.payload_offset() + RedirectBody::size_of()
+    }
+
+    #[inline]
+    fn data_len(&self) -> usize {
+        self.payload_len() - RedirectBody::size_of()
+    }
+
+    /// Returns the data packet as a `u8` slice
+    #[inline]
+    pub fn data(&self) -> &[u8] {
+        if let Ok(data) = self
+            .icmp()
+            .mbuf()
+            .read_data_slice(self.data_offset(), self.data_len())
+        {
+            unsafe { &*data.as_ptr() }
+        } else {
+            unreachable!()
```

Comment:
let's change this to `&[]` to avoid potentially panicking.

Reply:
good find. I know we moved that direction generally.

Reply:
:+1: